File: VB/CustomizationDemo/ShaderEffect.vb
Imports System
Imports System.Windows.Media.Effects
Imports System.Windows
Imports System.Windows.Media

Namespace CustomizationDemo

    Public Class GrayscaleEffect
        Inherits ShaderEffect

        Private Shared pixelShader As PixelShader = New PixelShader() With {.UriSource = New Uri("pack://application:,,,/CustomizationDemo;component/GrayscaleEffect.ps")}

        Public Shared ReadOnly InputProperty As DependencyProperty = RegisterPixelShaderSamplerProperty("Input", GetType(GrayscaleEffect), 0)

        Public Shared ReadOnly DesaturationFactorProperty As DependencyProperty = DependencyProperty.Register("DesaturationFactor", GetType(Double), GetType(GrayscaleEffect), New UIPropertyMetadata(0.0, PixelShaderConstantCallback(0), AddressOf CoerceDesaturationFactor))

        Private Shared Function CoerceDesaturationFactor(ByVal d As DependencyObject, ByVal value As Object) As Object
            Dim effect As GrayscaleEffect = CType(d, GrayscaleEffect)
            Dim newFactor As Double = CDbl(value)
            If newFactor < 0.0 OrElse newFactor > 1.0 Then Return effect.DesaturationFactor
            Return newFactor
        End Function

        Public Sub New()
            PixelShader = pixelShader
            UpdateShaderValue(InputProperty)
            UpdateShaderValue(DesaturationFactorProperty)
        End Sub

        Public Property Input As Brush
            Get
                Return CType(GetValue(InputProperty), Brush)
            End Get

            Set(ByVal value As Brush)
                SetValue(InputProperty, value)
            End Set
        End Property

        Public Property DesaturationFactor As Double
            Get
                Return CDbl(GetValue(DesaturationFactorProperty))
            End Get

            Set(ByVal value As Double)
                SetValue(DesaturationFactorProperty, value)
            End Set
        End Property
    End Class
End Namespace

File: CS/CustomizationDemo/ShaderEffect.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Windows.Media.Effects;
using System.Windows;
using System.Windows.Media;

namespace CustomizationDemo {
    public class GrayscaleEffect : ShaderEffect {
        static PixelShader pixelShader = new PixelShader() { 
            UriSource = new Uri(@"pack://application:,,,/CustomizationDemo;component/GrayscaleEffect.ps")
        };
        public static readonly DependencyProperty InputProperty = 
            ShaderEffect.RegisterPixelShaderSamplerProperty("Input", typeof(GrayscaleEffect), 0);
        public static readonly DependencyProperty DesaturationFactorProperty = 
            DependencyProperty.Register("DesaturationFactor", 
                                        typeof(double), 
                                        typeof(GrayscaleEffect),
                                        new UIPropertyMetadata(0.0, 
                                                               PixelShaderConstantCallback(0),
                                                               CoerceDesaturationFactor));
        static object CoerceDesaturationFactor(DependencyObject d, object value) {
            GrayscaleEffect effect = (GrayscaleEffect)d;
            double newFactor = (double)value;
            if(newFactor < 0.0 || newFactor > 1.0) 
                return effect.DesaturationFactor;
            return newFactor;
        }
        
        public GrayscaleEffect() {
            PixelShader = pixelShader;
            UpdateShaderValue(InputProperty);
            UpdateShaderValue(DesaturationFactorProperty);
        }
        public Brush Input {
            get { return (Brush)GetValue(InputProperty); }
            set { SetValue(InputProperty, value); }
        }
        public double DesaturationFactor {
            get { return (double)GetValue(DesaturationFactorProperty); }
            set { SetValue(DesaturationFactorProperty, value); }
        }
    }
}